File: sdks/javascript/src/openai.ts
// OpenAI SDK adapter for @wmcp/sdk.
import type { Tool } from "./index";

export interface OpenAITool {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: { type: "object"; properties: Record<string, any>; required?: string[] };
  };
}

export function toOpenAITools(tools: Tool[]): OpenAITool[] {
  return tools.map((t) => ({
    type: "function",
    function: {
      name: t.name,
      description: t.description,
      parameters:
        (t.inputSchema as OpenAITool["function"]["parameters"]) || {
          type: "object",
          properties: {},
        },
    },
  }));
}

/**
 * Run a `tool_calls[i].function` block from OpenAI's response via wmcp.
 */
export async function executeFunctionCall(
  client: { execute: (url: string, tool: string, args: any) => Promise<any> },
  url: string,
  functionCall: { name: string; arguments?: string | Record<string, any> }
) {
  let args: any = functionCall.arguments ?? {};
  if (typeof args === "string") {
    try {
      args = JSON.parse(args);
    } catch {
      args = {};
    }
  }
  return client.execute(url, functionCall.name, args);
}

File: sdks/python/wmcp/openai.py
"""OpenAI SDK adapter — turn wmcp Tools into the function-calling format."""

from __future__ import annotations

from typing import Any, Dict, List

from .client import Tool


def to_openai_tools(tools: List[Tool]) -> List[Dict[str, Any]]:
    """Convert wmcp tools to OpenAI's chat.completions tool format.

    Usage:

        from openai import OpenAI
        from wmcp import WmcpClient
        from wmcp.openai import to_openai_tools

        client = WmcpClient()
        wmcp_tools = client.tools("https://www.allbirds.com/products/mens-wool-runners")

        openai = OpenAI()
        resp = openai.chat.completions.create(
            model="gpt-4o",
            tools=to_openai_tools(wmcp_tools),
            messages=[{"role": "user", "content": "Find the cheapest size."}],
        )
    """
    return [
        {
            "type": "function",
            "function": {
                "name": t.name,
                "description": t.description,
                "parameters": t.input_schema,
            },
        }
        for t in tools
    ]


def execute_function_call(client, url: str, function_call: Dict[str, Any]) -> Dict[str, Any]:
    """Run an OpenAI `function_call` (or `tool_calls[i].function`) via wmcp.

    `function_call` is the dict with {"name": "...", "arguments": json_str_or_dict}.
    """
    import json as _json

    args = function_call.get("arguments", {})
    if isinstance(args, str):
        try:
            args = _json.loads(args)
        except _json.JSONDecodeError:
            args = {}
    return client.execute(url=url, tool=function_call["name"], args=args)
